### PR TITLE
Implement the `Fetcher` for V1Alpha contract

### PR DESCRIPTION
**Pull request by jamietanna**

### CODE REVIEW

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    var entity =
+        webClient
+            .get()
+            .uri(url)
+            .accept(MediaType.valueOf("application/vnd.uk.gov.api.v1alpha+json"))
+            .header(CORRELATION_ID_HEADER, correlationId)
+            .retrieve()
+            .onStatus(status(HttpStatus.NOT_ACCEPTABLE), handleNotAcceptable(url))
+            .onStatus(status(HttpStatus.BAD_REQUEST), handleBadRequest(url))
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:48]: Maybe?

```suggestion
            .onStatus(eq(HttpStatus.NOT_ACCEPTABLE), handleNotAcceptable(url))
            .onStatus(eq(HttpStatus.BAD_REQUEST), handleBadRequest(url))
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      logContractIssue(url);
+      return new ClientErrorException();
+    };
+  }
+
+  private static void logContractIssue(String url) {
+    LOGGER.error("Bad response body returned by {}, could be contract issue?", url);
+  }
+
+  private static Predicate<HttpStatus> status(HttpStatus status) {
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:91]: as above, 
```suggestion
  private static Predicate<HttpStatus> eq(HttpStatus status) {
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                  mdc.put(
+                      "response.v1alpha.correlation-id",
+                      err.headers().header(CORRELATION_ID_HEADER).get(0));
+                  LOGGER.error(
+                      "Bad request returned by {}, could be contract issue? {error: `{}`, error-description: `{}`}",
+                      url,
+                      e.getError().value(),
+                      e.getErrorDescription());
+                  h.error(new ClientErrorException());
+                });
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:117]: This could also be extracted to a method, thoughts?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  private static Function<Throwable, Throwable> handleError(String url) {
+    return exception -> {
+      if ((exception instanceof TemporaryErrorException)
+          || (exception instanceof ClientErrorException)
+          || (exception instanceof VersionNotSupportedException)) {
+        return exception;
+      }
+      // this could be for other reasons?
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:81]: do we need to do anything here?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    return apis;
+  }
+
+  private static Function<Throwable, Throwable> handleError(String url) {
+    return exception -> {
+      if ((exception instanceof TemporaryErrorException)
+          || (exception instanceof ClientErrorException)
+          || (exception instanceof VersionNotSupportedException)) {
+        return exception;
+      }
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:80]: Is it clear why we're doing this?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    mdc.put("request.v1alpha.correlation-id", correlationId);
+
+    var entity =
+        webClient
+            .get()
+            .uri(url)
+            .accept(MediaType.valueOf("application/vnd.uk.gov.api.v1alpha+json"))
+            .header(CORRELATION_ID_HEADER, correlationId)
+            .retrieve()
+            .onStatus(status(HttpStatus.NOT_ACCEPTABLE), handleNotAcceptable(url))
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:47]: This isn't called out on the OpenAPI - let's add it

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+import uk.gov.api.springboot.infrastructure.mappers.V1AlphaMapper;
+import uk.gov.api.springboot.infrastructure.models.metadata.v1alpha.ApiMetadata;
+import uk.gov.api.springboot.infrastructure.models.metadata.v1alpha.Data;
+
+class V1AlphaFetcherIntegrationTest {
+  @ExtendWith(SpringExtension.class)
+  @Import({IntegrationTestBase.Config.class, JacksonAutoConfiguration.class})
+  @DirtiesContext
+  @Nested
+  abstract class IntegrationTestBase {
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:50]: is it clear why we've done this?

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+
+    protected String baseUrl() {
+      return "%s/apis".formatted(server.url("").toString());
+    }
+  }
+
+  @Nested
+  class Fetch {
+    @Nested
+    class ___fTODO extends IntegrationTestBase { // TODO
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:138]: needs renaming

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+
+      @Test
+      void delegates__() {
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:141]: needs renaming

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+
+      @Test
+      void itLogsNumberOfItemsFetched() {
+        server.enqueue(success());
+
+        fetch();
+
+        assertThat(logger)
+            .anyThread()
+            .hasLogged(LoggingEvent.info("Successfully retrieved {} APIs from {}", 0, baseUrl()));
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:226]: do we want to add some valid options here to make sure it works?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    this.mdc = mdc;
+  }
+
+  @Override
+  public List<Api> fetch(String baseUrl) {
+    String url = baseUrl + "/apis";
+    String correlationId = mdc.get(MdcFacade.CORRELATION_ID);
+    mdc.put("request.v1alpha.correlation-id", correlationId);
+
+    var entity =
```
Comment (nahteb) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:40]: I think I'd prefer to have the type here rather than `var`, it's a bit hard to see what it is

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        return exception;
+      }
+      // this could be for other reasons?
+      logContractIssue(url);
+      return new ClientErrorException();
+    };
+  }
+
+  private static void logContractIssue(String url) {
+    LOGGER.error("Bad response body returned by {}, could be contract issue?", url);
```
[the remark this answers is not in the document]
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:88]: I think error makes sense

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java (hunk 7 above)
Comment (nahteb) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:50]: yes makes sense

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java (hunk 5 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:80]: yes makes sense

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java (hunk 6 above)
Comment (jamietanna) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:47]: Follow-up in https://github.com/co-cddo/federated-api-model/issues/129

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            .block();
+    if (entity == null) {
+      logContractIssue(url);
+      throw new ClientErrorException();
+    }
+    var body = entity.getBody();
+    if (body == null) {
+      logContractIssue(url);
+      throw new ClientErrorException();
+    }
```
[the remark this answers is not in the document]
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:64]: leave it as in not have a test for it? Or as in just log it?

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java (hunk 4 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:84]: I can't think of anything else

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java (hunk 10 above)
Comment (nahteb) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:226]: yeah sounds good

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java (hunk 1 above)
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:51]: `eq` makes sense to me

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jamietanna) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:64]: Sorry, "leave it" here being "don't add the test coverage for the case that these are `null`, as it'd require a unit test, and we're only set up for unit integration right now"

[the remark this answers is not in the document]
Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:64]: ah ok, yep that makes sense

**Comment by jamietanna** [comment]
Also verified this works through:

```java

  @Test
  void t() {
    var b = WebClient.builder().build();
    V1AlphaFetcher fetcher = new V1AlphaFetcher(b, new V1AlphaMapper(), new MdcFacade());
    var out =
        fetcher.fetch("https://federated-api-model-spring-boot-sandbox.london.cloudapps.digital");

    System.out.println(out);
  }

```

Which outputs:

```
[Api[name=Federated API Model Sandbox, description=A sample deployment of the Federated API Model's reference implementation, url=https://federated-api-model-spring-boot-sandbox.london.cloudapps.digital, contact=api-programme@digital.cabinet-office.gov.uk, organisation=CDDO, documentationUrl=https://github.com/co-cddo/federated-api-model]]
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+    class InGeneral extends IntegrationTestBase {
+
+      @Test
+      void delegatesToMapperFromMethod() {
+        var api = fakeApi();
+        var expected = new ApiMetadata();
+        expected.setApiVersion(ApiMetadata.ApiVersion.API_GOV_UK_V_1_ALPHA);
+        var data = new Data();
+        data.setName("T");
+        data.setName("The API Name");
```
Comment (nahteb) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcherIntegrationTest.java:147]:
```suggestion
        data.setName("The API Name");
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    mdc.put("request.v1alpha.correlation-id", correlationId);
+
+    ResponseEntity<BulkMetadataResponse> entity =
+        webClient
+            .get()
+            .uri(url)
+            .accept(MediaType.valueOf("application/vnd.uk.gov.api.v1alpha+json"))
+            .header(CORRELATION_ID_HEADER, correlationId)
+            .retrieve()
+            .onStatus(eq(HttpStatus.NOT_FOUND), handleNotFound(url))
```
Comment (jamietanna) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:50]: did we want this in the contract? It feels a bit weird saying it. Maybe one more for the specification for "how fetchers should work"

Comment (nahteb) [reply examples/java/spring-boot/src/main/java/uk/gov/api/springboot/infrastructure/fetcher/v1alpha/V1AlphaFetcher.java:50]: having it in the spec instead makes sense to me

**Review by nahteb — APPROVED** [review]
